Improve how `mypy` is run

llucax opened the pull request:
This PR mainly split `mypy` check for the source package and dev files.

`mypy` tends to have issues with checking source files inside a `src/` directory. There are tricks[^1] but it seems like the only way to do it is to run `mypy` on `.` using `MYPYPATH=src`, which will still check the whole `.` directory (include whole virtualenvs if there are any), which is not what we want. For now we've been using `-p` to check for packages instead of files, which seemed to have done the trick, but we need to pass directories that don't really have a package structure (like benchmarks or examples or tests) as packages, which is also not great.

We try a different approach, we check the source code as a package separately, and then check the development files (tests, examples, etc.), which are passed as simple paths. This also opens up the door to using more relaxed rules for these development files if needed in the future.

To achieve this we also move the `packages` specification to the `pyproject.toml` file, so now `mypy` can be called without any arguments to check the source code package, making it easier to call it without relaying in `nox`.

We also move all the options to the `pyproject.toml` file, so `mypy` can be run consistently also outside of `nox`, and the stub dependencies now need to be specified manually in the `pyproject.toml` file, so they can be pinned.

This is also a step towards removing `nox` (#71).

[^1]: https://mypy.readthedocs.io/en/stable/running_mypy.html#mapping-file-paths-to-modules

llucax committed:
Move options to pyproject.toml

It is possible to define most options within the `pyproject.toml` file,
which offers several advantages: the tool can be used from the command
line directly (bypassing the need for `nox`), without having to remember
the options. It also facilitates other integrations that need to invoke
mypy directly.

Signed-off-by: Leandro Lucarella <luca-frequenz@llucax.com>

llucax committed:
Split `mypy` check for the source package and dev files

`mypy` tends to have issues with checking source files inside a `src/`
directory. There are tricks[1] but it seems like the only way to do it
is to run `mypy` on `.` using `MYPYPATH=src`, which will still check the
whole `.` directory (include whole virtualenvs if there are any), which
is not what we want. For now we've been using `-p` to check for packages
instead of files, which seemed to have done the trick, but we need to
pass directories that don't really have a package structure (like
benchmarks or examples or tests) as packages, which is also not great.

This commit tries a different approach, we check the source code as a
package separately, and then check the development files (tests,
examples, etc.), which are passed as simple paths. This also opens up
the door to using more relaxed rules for these development files if
needed in the future.

To achieve this we also move the `packages` specification to the
`pyproject.toml` file, so now `mypy` can be called without any arguments
to check the source code package, making it easier to call it without
relaying in `nox`.

[1] https://mypy.readthedocs.io/en/stable/running_mypy.html#mapping-file-paths-to-modules

Signed-off-by: Leandro Lucarella <luca-frequenz@llucax.com>

llucax committed:
Remove the (now) unused `Config.package_args()` method

This was used by the `mypy` `nox` session.

Signed-off-by: Leandro Lucarella <luca-frequenz@llucax.com>

llucax committed:
Remove old `pylint` comment about the SDK

Signed-off-by: Leandro Lucarella <luca-frequenz@llucax.com>

llucax committed:
Don't install stubs automatically with `mypy`

When installing stubs automatically with `mypy` we can't pin the
dependencies, and we make it harder to just run `mypy` without going
through `nox`.

From now on, dependencies must be explicitly declared in the
`pyproject.toml` file, in the optional `dev-mypy` dependencies.

Signed-off-by: Leandro Lucarella <luca-frequenz@llucax.com>

Marenz commented:
First commit has a weird phrasing "Most options can be specified in the pyproject.toml file **and this have some** advantages."

Marenz commented on pyproject.toml:
```diff
@@ -79,7 +79,11 @@ dev-mkdocs = [
   "mypy == 1.5.1",
-  "types-setuptools >= 67.6.0, < 68", # Should match the global dependency
+  "types-setuptools == 68.1.0.0", # Should match the build dependency
+  "types-Markdown == 3.4.2.10",
```
How is this updated when the mypy version changes? Like, who keeps it in sync? Dependabot would just pick the latest versions no? And if those don't match what mypy deps need?

llucax replied:
Mypy is not in charge of these dependencies, they usually need to match the python package, like `types-markdown` should have the sameish version as the `markdown` package that we are using.

We are actually not pinning transitional dependencies, which at some point could be problematic (this is a problem to think about in the future, there are many optional *package managers* that take care of this, like `pipenv`). For example, in this case we don't depend directly on the `markdown` package, so it is a bit strange that we need the stub.

All that mypy does is somehow looking for packages without type info and look for them in https://github.com/python/typeshed to see if there is a *stub* package providing it and then installing that via `pip`.

Marenz commented:
two comments, otherwise LGTM

llucax committed:
Add release notes

Signed-off-by: Leandro Lucarella <luca-frequenz@llucax.com>

llucax committed:
Add summary and clean up release notes

This is to prepare for the release.

Signed-off-by: Leandro Lucarella <luca-frequenz@llucax.com>

llucax committed:
Improve `darglint` related upgrading notes

Signed-off-by: Leandro Lucarella <luca-frequenz@llucax.com>

llucax committed:
Add `mypy` no-incremental option as a comment

This option avoids using a cache. It is slower but prevents some issues
with `mypy` giving different results on different runs (typically
complaining about some `type: ignore[issue]` not being used but when
removing it complaining about an `issue` again.

It is included so it has more visibility to users in the hope that they
can spend less time debugging if they hit this `mypy` issue.

Signed-off-by: Leandro Lucarella <luca-frequenz@llucax.com>

llucax commented:
Rephrased the commit, squashed the fixup commits and added a few more commits, most notably release notes and the `no_incremental` option. Testing via https://github.com/frequenz-floss/frequenz-sdk-python/pull/587.

llucax commented:
Not tested in https://github.com/frequenz-floss/frequenz-sdk-python/pull/587 anymore because all the `pydoclint` tests are too disruptive. Waiting to test it after that PR (which will include an ad-hoc hack) is merged.

llucax commented:
The fix for `mypy` weird behavior is to disable caching, which makes it slower (although not by a **LOT**). I think all the other changes here are still worth it, so I would keep them, but I'm not sure if we should disable caching by default or not, as it is a very awful problem to debug but probably not affecting other simpler project (specially ones not using grpc stuff), so maybe slow everything down just in case is not justified. I'm considering adding the option to the template but commented out, with a comment explaining why is it there.

llucax commented:
New PR for testing this: https://github.com/frequenz-floss/frequenz-sdk-python/pull/620

llucax committed:
Disable some pylint checks that are also checked by flake8

Flake8 is faster, so we prefer it.

Signed-off-by: Leandro Lucarella <luca-frequenz@llucax.com>

llucax commented:
The test PR seems to be working well, so this is ready for the final review and merge (and release).

llucax commented:
Enabling auto-merge.

llucax committed:
Move `flake8` earlier in the `ci_checks_max` session

We want to run the checks from the fatests to the slowest, so we get
errors as early as possible.

Signed-off-by: Leandro Lucarella <luca-frequenz@llucax.com>

llucax committed:
Bump pydoclint to 0.3.1

This new release will not require a function `yield`ing `None` to have
a `Yields:` section.

Signed-off-by: Leandro Lucarella <luca-frequenz@llucax.com>
